src/main/java/com/google/sps/servlets/Caption.java

```diff
@@ -0,0 +1,53 @@
+package com.google.sps.servlets;
+
+import java.io.IOException;
+import java.io.InputStream;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import org.w3c.dom.Document;
+import org.w3c.dom.NodeList;
+import org.xml.sax.SAXException;
+
+public class Caption {
+  /*
+   * Usage: String caption = new Caption().getCaptionFromID("abcdefg");
+   * 
+   * Return null if not found.
+   */
+
+  public CaptionService captionService;
+
+  public Caption() {
+    captionService = new CaptionService();
+  }
+
+  public void setCaptionService(CaptionService captionService) {
+    this.captionService = captionService;
+  }
+
+  public String getCaptionFromId(String videoId) {
+    // BCP47 language tags
+    final String[] languages = {"en", "en-US", "en-GB", "en-AU", "en-IE", "en-ZA"};
+    for (String language : languages) {
+      try {
+        return parseXmlFromStream(captionService.fetchStream(videoId, language));
+      } catch (SAXException | ParserConfigurationException | IOException
+          | IllegalArgumentException e) {
+        continue;
+      }
+    }
+    return null;
+  }
+
+  private String parseXmlFromStream(InputStream inputStream)
+      throws SAXException, ParserConfigurationException, IOException, IllegalArgumentException {
+    Document doc = DocumentBuilderFactory.newInstance().newDocumentBuilder().parse(inputStream);
+    NodeList textDom = doc.getElementsByTagName("text");
+    String caption = "";
+    int lenTextDom = textDom.getLength();
+    for (int i = 0; i < lenTextDom; i++) {
+      caption = caption + " " + textDom.item(i).getFirstChild().getNodeValue();
+    }
```

Comment:
Algorithmic complexity here is O(|caption|^2). Java Strings are immutable and when you concatenate two strings, it is actually building a new string from scratch taking characters from both input strings. Instead of concatenating a new strings you want to append new characters into an existing mutable string object. StringBuilder is a good candidate. 

FYI: This is a common mistake in Java/Python interviews. C++ concatenating with "+" creates a new string so it has the same issue but using "+=" actually appends to the string efficiently because C++ strings are mutable.

Reply:
Done. Good catch! I guess this is a pitfall that I should be very careful about from now on